src/rust-crypto/RoomEncryptor.ts

```diff
@@ -118,11 +121,45 @@ export class RoomEncryptor {
      * @param globalBlacklistUnverifiedDevices - When `true`, it will not send encrypted messages to unverified devices
      */
     public async prepareForEncryption(globalBlacklistUnverifiedDevices: boolean): Promise<void> {
-        const logger = new LogSpan(this.prefixedLogger, "prepareForEncryption");
+        // We consider a prepareForEncryption as an encryption promise as it will potentially share keys
+        // even if it doesn't send an event.
+        // Usually this is called when the user starts typing, so we want to make sure we have keys ready when the
+        // message is finally sent. The actual event encryption request will arrive after wait for the prepareForEncryption
+        // promise to resolve, and then do again an ensureEncryptionSession that should be no op as we already share the room key.
```

Comment:
> The actual event encryption request will arrive after wait for the prepareForEncryption promise to resolve

I don't really understand this, or what it is trying to say. "The actual event encryption request" == "a call to `encryptEvent`"?  So you're saying that the application will wait for `prepareForEncryption` to complete before it calls `encryptEvent`? I don't think that's right.

Reply:
I think it's how it's supposed to work. Legacy is doing the same.
If you starts typing then call encrypt event, we let the prepare finish, then do the encrypt. The share room key part for the secong call should be no op as the room key is already shared

Reply:
Notice that anyhow the olmMachine.shareRoomKey() should be locked per room

Reply:
> if you starts typing then call encrypt event, we let the prepare finish, then do the encrypt.

but that's not what the comment says?

Reply:
... or rather the comment says that we expect the application to wait, and I don't think we do. That's why we have to wait here.

Reply:
I updated the comment:
```
// If `encryptEvent` is invoked before `prepareForEncryption` has completed, the `encryptEvent` call will wait for
// `prepareForEncryption` to complete before executing.
// The part where `encryptEvent` shares the room key would be no op as it was already performed by prepareForEncryption.